### PR TITLE
add rowset state

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. add rowset state to rowset
2. add close api to rowset to release resources
issue: #2665

### CODE REVIEW

File: be/src/olap/rowset/rowset.h [hunk 1]
```diff
@@ -152,8 +268,13 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
     bool _is_pending;    // rowset is pending iff it's not in visible state
     bool _is_cumulative; // rowset is cumulative iff it's visible and start version < end version
 
-    DorisCallOnce<OLAPStatus> _load_once;
+    SpinLock _lock;
+    Mutex _load_lock;
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: give comment for the usage of these two lock

File: be/src/olap/rowset/rowset.h [hunk 2]
```diff
@@ -152,8 +268,13 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
...
     bool _need_delete_file = false;
+    // variable to indicate how many rowset readers owned this rowset
+    std::atomic<uint64_t> _refs_by_reader;
+    // rowset state machine
+    RowsetStateMachine _rowset_state_machine;
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: should be given a default value

File: be/src/olap/rowset/rowset.h [hunk 3]
```diff
@@ -95,6 +151,37 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
...
+        {
+            std::lock_guard<SpinLock> l(_lock);
+            old_state = _rowset_state_machine.rowset_state();
+            if (old_state == ROWSET_UNLOADED) {
+                return;
+            }
+            uint64_t current_refs = _refs_by_reader;
+            st = _rowset_state_machine.on_close(current_refs);
+            if (current_refs == 0) {
+                do_close();
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: better to call this out of lock block

File: be/src/olap/rowset/rowset.h [hunk 4]
```diff
@@ -127,6 +214,30 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
...
+    void aquire() {
+        ++_refs_by_reader;
+    }
+
+    void release() {
+        // if the refs by reader is 0 and the rowset is closed, should release the resouce
+        uint64_t current_refs = --_refs_by_reader;
+        if (current_refs == 0 && _rowset_state_machine.rowset_state() == ROWSET_UNLOADING) {
+            auto st = _rowset_state_machine.on_close(current_refs);
+            if (st != OLAP_SUCCESS) {
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: seems no usage, just leave a DCHECK(state == UNLOADED)

File: be/src/olap/rowset/rowset.h [hunk 5]
```diff
@@ -95,6 +151,37 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
     // TODO should we rename the method to remove_files() to be more specific?
     virtual OLAPStatus remove() = 0;
 
+    // close to clear the resource owned by rowset
+    // including: open files, indexes and so on
+    // NOTICE: can not call this function in multithreads
+    void close() {
+        RowsetState old_state = _rowset_state_machine.rowset_state();
+        if (old_state == ROWSET_UNLOADED) {
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: When the state is unloading, seems OK to return

File: be/src/olap/rowset/rowset.h [hunk 6]
```diff
@@ -95,6 +151,37 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
...
+            return;
+        }
+        OLAPStatus st = OLAP_SUCCESS;
+        {
+            std::lock_guard<SpinLock> l(_lock);
+            old_state = _rowset_state_machine.rowset_state();
+            if (old_state == ROWSET_UNLOADED) {
```
Comment (imay) [inline be/src/olap/rowset/rowset.h]: UNLOADING

File: be/src/olap/rowset/rowset.cpp [hunk 7]
```diff
@@ -36,7 +37,34 @@ Rowset::Rowset(const TabletSchema *schema,
...
+        load_log = ss.str();
+    }
+    if (load_log != "") {
+        LOG(INFO) << load_log;
+    }
+    return OLAP_SUCCESS;
+}
+
+OLAPStatus Rowset::create_reader(std::shared_ptr<RowsetReader>* result) {
+    std::lock_guard<SpinLock> l(_lock);
```
Comment (imay) [inline be/src/olap/rowset/rowset.cpp]: Is this lock needed?

File: be/src/olap/rowset/rowset.h (hunk 1 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset.h]: done

File: be/src/olap/rowset/rowset.h (hunk 5 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset.h]: yes, u are right.

File: be/src/olap/rowset/rowset.cpp (hunk 7 above)
Comment (kangpinghuang) [reply be/src/olap/rowset/rowset.cpp]: useless

File: be/src/olap/rowset/alpha_rowset.cpp [hunk 8]
```diff
@@ -51,7 +51,8 @@ OLAPStatus AlphaRowset::do_load(bool use_cache) {
     return OLAP_SUCCESS;
 }
 
-OLAPStatus AlphaRowset::do_create_reader(std::shared_ptr<RowsetReader>* result) {
+OLAPStatus AlphaRowset::create_reader(std::shared_ptr<RowsetReader>* result) {
+    MutexLock release_lock(&_lock);
```
Comment (imay) [inline be/src/olap/rowset/alpha_rowset.cpp]: why lock? I think you should call load to make sure rowset is loaded?

Comment (kangpinghuang) [reply be/src/olap/rowset/alpha_rowset.cpp]: yes, there should be no lock here, I will remove it.
load is called in init()

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**